### ISSUE TITLE
Question: is Yas thread-safe?

### ISSUE THREAD

**Issue by Hi-Angel** [niXman/yas#34]
I know, sounds like a silly question — like, why would one make a non thread-safe serialization lib — but e.g. [Cereal turned out](https://github.com/USCiLab/cereal/pull/498) to be one. It became less funny, when I looked at Boost::serialization, from which Cereal borrowed the API, and found out neither it is a thread-safe library! Sick!

So, I'm just trying to be on the safe side: is Yas library thread-safe?

**Comment by niXman** [comment]
Hi,

> is Yas library thread-safe?

No.

**Comment by niXman** [comment]
Hi,

Bring specific situations which you are interested, 
Please provide the specific situations you are interested in, otherwise it is a very broad topic... Looking ahead, I note - YAS is state-less, except i/o-streams.

**Comment by huyuguang** [comment]
To honest, before I saw this post, I think all serialization libs are thread-safe. So I am very shocked when I saw this post.
As my understand, when talking about thread-safe, we need more details, for example, the STL requires:

> implementations are required to avoid data races when the contents of the contained object in different elements in the same sequence, excepting vector<bool>, are modified concurrently.

So, vector_a.push_back() in two threads is NOT safe, vector_a.push_back() in one thread and vector_b.push_back() in another thread is safe, vector_a[0].func() in one thread and vector_a[1].func() in another thread is safe.

About YAS, I want to know the what operation is safe and what is not.
